modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeTopologyService.java

```diff
@@ -52,31 +52,50 @@ void setLocalMember(Member member) {
      */
     void onMembershipEvent(MembershipEvent event) {
         ClusterNode member = fromMember(event.member());
-        for (TopologyEventHandler handler : getEventHandlers()) {
-            switch (event.type()) {
-                case ADDED:
-                    members.put(member.name(), member);
 
-                    handler.onAppeared(member);
+        String memberName = member.name();
+
+        boolean appeared = false;
+        boolean disappeared = false;
+
+        switch (event.type()) {
+            case ADDED:
+                members.put(memberName, member);
+
+                appeared = true;
+
+                break;
 
-                    break;
+            case LEAVING:
+                members.remove(memberName);
 
-                case LEAVING:
-                    members.remove(member.name());
+                disappeared = true;
 
-                    handler.onDisappeared(member);
+                break;
 
-                    break;
+            case REMOVED:
+                if (members.containsKey(memberName)) {
+                    members.remove(memberName);
 
-                case REMOVED:
-                case UPDATED:
-                    // No-op.
-                    break;
+                    disappeared = true;
+                }
 
-                default:
-                    throw new IgniteInternalException("This event is not supported: event = " + event);
+                break;
 
-            }
+            case UPDATED:
+                // No-op.
+                break;
+
+            default:
+                throw new IgniteInternalException("This event is not supported: event = " + event);
+
+        }
+
+        for (TopologyEventHandler handler : getEventHandlers()) {
```

Comment:
We will re-iterate over all event handlers even if we've received a non-interesting message (e.g. `UPDATED`). I think we can optimize this approach. For example, I would suggest using two methods (`fireAppearedEvent` and `fireDissapearedEvent`) and get rid of these two flags